wgpu-core/src/device.rs

```diff
@@ -596,16 +600,16 @@ impl<B: GfxBackend> Device<B> {
             &self.desc_allocator,
             force_wait,
         );
+        {
+            let mut last_done_guard = self.last_done.lock();
+            *last_done_guard = last_done_guard.max(last_done);
```

Comment:
I don't think we need a `Mutex` here. After this function is called in `submit`, the device is re-locked for writing, so we can just assign the value in there directly.

Reply:
`device.pending` is locked but the device itself isn't given a write a lock. I could introduce a write lock between the `maintain` call and further uses of the device, if that works better.

Reply:
Ok, I think I know what we can do to avoid this field. At any point, we can consider `active[0].index` to be the first *undone* submission index. So we can take that and pass it to the `pool.maintain()` as a different semantics (first undone), and in side the `maintain()` implementation just change the `<= last_done` comparison to `< first_undone`. How does that sound?

Reply:
Sounds good to me, seems straightforward to implement as well.